File: services/ansible-service/internal/parser/playbook_parser.py
"""Parser YAML pour analyser les playbooks Ansible."""
import logging
from typing import Optional, List, Dict, Any
import yaml
from pathlib import Path

logger = logging.getLogger(__name__)


class PlaybookParser:
    """Parser pour analyser les playbooks Ansible en profondeur."""

    @staticmethod
    def parse_playbook(content: str) -> Optional[Dict[str, Any]]:
        """Parse le contenu YAML d'un playbook."""
        try:
            data = yaml.safe_load(content)
            if not data:
                return None

            # Un playbook peut être une liste de plays ou un seul play
            plays = data if isinstance(data, list) else [data]

            parsed_plays = []
            for play in plays:
                parsed_play = PlaybookParser._parse_play(play)
                if parsed_play:
                    parsed_plays.append(parsed_play)

            return {
                "plays": parsed_plays,
                "play_count": len(parsed_plays),
            }
        except yaml.YAMLError as e:
            logger.error(f"Erreur lors du parsing YAML: {e}")
            return None
        except Exception as e:
            logger.error(f"Erreur lors de l'analyse du playbook: {e}")
            return None

    @staticmethod
    def _parse_play(play: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Parse un play individuel."""
        if not isinstance(play, dict):
            return None

        parsed = {
            "name": play.get("name", "Unnamed play"),
            "hosts": play.get("hosts", "all"),
            "gather_facts": play.get("gather_facts", True),
            "vars": play.get("vars", {}),
            "vars_files": play.get("vars_files", []),
            "vars_prompt": play.get("vars_prompt", []),
            "pre_tasks": [],
            "tasks": [],
            "handlers": [],
            "post_tasks": [],
            "roles": [],
            "collections": play.get("collections", []),
            "become": play.get("become", False),
            "become_user": play.get("become_user"),
            "become_method": play.get("become_method"),
            "environment": play.get("environment", {}),
        }

        # Parser les pre_tasks
        if "pre_tasks" in play:
            parsed["pre_tasks"] = PlaybookParser._parse_tasks(play["pre_tasks"])

        # Parser les tasks
        if "tasks" in play:
            parsed["tasks"] = PlaybookParser._parse_tasks(play["tasks"])

        # Parser les handlers
        if "handlers" in play:
            parsed["handlers"] = PlaybookParser._parse_tasks(play["handlers"])

        # Parser les post_tasks
        if "post_tasks" in play:
            parsed["post_tasks"] = PlaybookParser._parse_tasks(play["post_tasks"])

        # Parser les roles
        if "roles" in play:
            parsed["roles"] = PlaybookParser._parse_roles(play["roles"])

        return parsed

    @staticmethod
    def _parse_tasks(tasks: List[Any]) -> List[Dict[str, Any]]:
        """Parse une liste de tâches."""
        parsed_tasks = []
        for task in tasks:
            if isinstance(task, dict):
                parsed_task = PlaybookParser._parse_task(task)
                if parsed_task:
                    parsed_tasks.append(parsed_task)
            elif isinstance(task, str):
                # Tâche inline simple
                parsed_tasks.append({
                    "name": task,
                    "type": "inline",
                })
        return parsed_tasks

    @staticmethod
    def _parse_task(task: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Parse une tâche individuelle."""
        parsed = {
            "name": task.get("name", "Unnamed task"),
            "module": None,
            "args": {},
            "when": task.get("when"),
            "loop": task.get("loop"),
            "until": task.get("until"),
            "retries": task.get("retries"),
            "delay": task.get("delay"),
            "register": task.get("register"),
            "changed_when": task.get("changed_when"),
            "failed_when": task.get("failed_when"),
            "ignore_errors": task.get("ignore_errors", False),
            "async_val": task.get("async"),
            "poll": task.get("poll"),
            "delegate_to": task.get("delegate_to"),
            "delegate_facts": task.get("delegate_facts", False),
            "become": task.get("become"),
            "become_user": task.get("become_user"),
            "environment": task.get("environment", {}),
            "tags": task.get("tags", []),
            "notify": task.get("notify", []),
        }

        # Identifier le module utilisé
        # Les modules peuvent être définis comme clés directes ou via "action"
        module_keys = [k for k in task.keys() if k not in [
            "name", "when", "loop", "until", "retries", "delay", "register",
            "changed_when", "failed_when", "ignore_errors", "async", "poll",
            "delegate_to", "delegate_facts", "become", "become_user",
            "environment", "tags", "notify", "action", "local_action",
            "with_", "vars", "include", "include_tasks", "import_tasks",
            "block", "rescue", "always", "role", "include_role", "import_role",
        ]]

        if module_keys:
            parsed["module"] = module_keys[0]
            parsed["args"] = task.get(module_keys[0], {})
        elif "action" in task:
            action = task["action"]
            if isinstance(action, dict):
                parsed["module"] = list(action.keys())[0] if action else None
                parsed["args"] = list(action.values())[0] if action else {}
            elif isinstance(action, str):
                parsed["module"] = action.split()[0] if action else None

        # Gérer les includes et imports
        if "include" in task:
            parsed["type"] = "include"
            parsed["include"] = task["include"]
        elif "include_tasks" in task:
            parsed["type"] = "include_tasks"
            parsed["include_tasks"] = task["include_tasks"]
        elif "import_tasks" in task:
            parsed["type"] = "import_tasks"
            parsed["import_tasks"] = task["import_tasks"]
        elif "include_role" in task:
            parsed["type"] = "include_role"
            parsed["include_role"] = task["include_role"]
        elif "import_role" in task:
            parsed["type"] = "import_role"
            parsed["import_role"] = task["import_role"]
        elif "role" in task:
            parsed["type"] = "role"
            parsed["role"] = task["role"]
        elif "block" in task:
            parsed["type"] = "block"
            parsed["block"] = PlaybookParser._parse_tasks(task.get("block", []))
            parsed["rescue"] = PlaybookParser._parse_tasks(task.get("rescue", []))
            parsed["always"] = PlaybookParser._parse_tasks(task.get("always", []))
        else:
            parsed["type"] = "task"

        return parsed

    @staticmethod
    def _parse_roles(roles: List[Any]) -> List[Dict[str, Any]]:
        """Parse une liste de rôles."""
        parsed_roles = []
        for role in roles:
            if isinstance(role, str):
                parsed_roles.append({
                    "name": role,
                    "vars": {},
                })
            elif isinstance(role, dict):
                parsed_roles.append({
                    "name": role.get("role", role.get("name", "unknown")),
                    "vars": {k: v for k, v in role.items() if k != "role" and k != "name"},
                })
        return parsed_roles

    @staticmethod
    def analyze_playbook(content: str) -> Dict[str, Any]:
        """Analyse complète d'un playbook avec statistiques."""
        parsed = PlaybookParser.parse_playbook(content)
        if not parsed:
            return {"error": "Impossible de parser le playbook"}

        # Calculer les statistiques
        stats = {
            "total_plays": parsed["play_count"],
            "total_tasks": 0,
            "total_handlers": 0,
            "total_pre_tasks": 0,
            "total_post_tasks": 0,
            "total_roles": 0,
            "modules_used": set(),
            "hosts_targeted": set(),
            "become_used": False,
            "collections_used": set(),
        }

        for play in parsed["plays"]:
            stats["total_tasks"] += len(play.get("tasks", []))
            stats["total_handlers"] += len(play.get("handlers", []))
            stats["total_pre_tasks"] += len(play.get("pre_tasks", []))
            stats["total_post_tasks"] += len(play.get("post_tasks", []))
            stats["total_roles"] += len(play.get("roles", []))

            hosts = play.get("hosts")
            if isinstance(hosts, str):
                stats["hosts_targeted"].add(hosts)
            elif isinstance(hosts, list):
                stats["hosts_targeted"].update(hosts)

            if play.get("become"):
                stats["become_used"] = True

            stats["collections_used"].update(play.get("collections", []))

            # Extraire les modules utilisés
            for task_list in [play.get("tasks", []), play.get("handlers", []), 
                            play.get("pre_tasks", []), play.get("post_tasks", [])]:
                for task in task_list:
                    if task.get("module"):
                        stats["modules_used"].add(task["module"])

        return {
            "parsed": parsed,
            "statistics": {
                "total_plays": stats["total_plays"],
                "total_tasks": stats["total_tasks"],
                "total_handlers": stats["total_handlers"],
                "total_pre_tasks": stats["total_pre_tasks"],
                "total_post_tasks": stats["total_post_tasks"],
                "total_roles": stats["total_roles"],
                "modules_used": sorted(list(stats["modules_used"])),
                "hosts_targeted": sorted(list(stats["hosts_targeted"])),
                "become_used": stats["become_used"],
                "collections_used": sorted(list(stats["collections_used"])),
            },
        }

File: services/ansible-service/internal/parser/playbook_parser.go
// Package parser analyse les playbooks Ansible (YAML) : structure des plays,
// tâches, rôles, et statistiques agrégées.
package parser

import (
	"sort"

	"gopkg.in/yaml.v3"
)

// taskReservedKeys sont les clés d'une tâche qui ne désignent pas un module.
var taskReservedKeys = map[string]bool{
	"name": true, "when": true, "loop": true, "until": true, "retries": true,
	"delay": true, "register": true, "changed_when": true, "failed_when": true,
	"ignore_errors": true, "async": true, "poll": true, "delegate_to": true,
	"delegate_facts": true, "become": true, "become_user": true,
	"environment": true, "tags": true, "notify": true, "action": true,
	"local_action": true, "with_": true, "vars": true, "include": true,
	"include_tasks": true, "import_tasks": true, "block": true, "rescue": true,
	"always": true, "role": true, "include_role": true, "import_role": true,
}

// ParsePlaybook parse le contenu YAML d'un playbook. Retourne nil si le
// contenu est vide ou invalide.
func ParsePlaybook(content string) map[string]any {
	var root any
	if err := yaml.Unmarshal([]byte(content), &root); err != nil || root == nil {
		return nil
	}

	plays, ok := root.([]any)
	if !ok {
		plays = []any{root}
	}

	parsedPlays := []map[string]any{}
	for _, play := range plays {
		if parsed := parsePlay(play); parsed != nil {
			parsedPlays = append(parsedPlays, parsed)
		}
	}

	return map[string]any{
		"plays":      parsedPlays,
		"play_count": len(parsedPlays),
	}
}

func parsePlay(raw any) map[string]any {
	play, ok := toStringMap(raw)
	if !ok {
		return nil
	}

	parsed := map[string]any{
		"name":          getOr(play, "name", "Unnamed play"),
		"hosts":         getOr(play, "hosts", "all"),
		"gather_facts":  getOr(play, "gather_facts", true),
		"vars":          getOr(play, "vars", map[string]any{}),
		"vars_files":    getOr(play, "vars_files", []any{}),
		"vars_prompt":   getOr(play, "vars_prompt", []any{}),
		"pre_tasks":     []map[string]any{},
		"tasks":         []map[string]any{},
		"handlers":      []map[string]any{},
		"post_tasks":    []map[string]any{},
		"roles":         []map[string]any{},
		"collections":   getOr(play, "collections", []any{}),
		"become":        getOr(play, "become", false),
		"become_user":   play["become_user"],
		"become_method": play["become_method"],
		"environment":   getOr(play, "environment", map[string]any{}),
	}

	for _, section := range []string{"pre_tasks", "tasks", "handlers", "post_tasks"} {
		if raw, found := play[section]; found {
			parsed[section] = parseTasks(raw)
		}
	}
	if roles, found := play["roles"]; found {
		parsed["roles"] = parseRoles(roles)
	}

	return parsed
}

func parseTasks(raw any) []map[string]any {
	list, ok := raw.([]any)
	if !ok {
		return []map[string]any{}
	}
	parsed := []map[string]any{}
	for _, item := range list {
		switch task := item.(type) {
		case map[string]any:
			if p := parseTask(task); p != nil {
				parsed = append(parsed, p)
			}
		case string:
			parsed = append(parsed, map[string]any{"name": task, "type": "inline"})
		}
	}
	return parsed
}

func parseTask(task map[string]any) map[string]any {
	parsed := map[string]any{
		"name":           getOr(task, "name", "Unnamed task"),
		"module":         nil,
		"args":           map[string]any{},
		"when":           task["when"],
		"loop":           task["loop"],
		"until":          task["until"],
		"retries":        task["retries"],
		"delay":          task["delay"],
		"register":       task["register"],
		"changed_when":   task["changed_when"],
		"failed_when":    task["failed_when"],
		"ignore_errors":  getOr(task, "ignore_errors", false),
		"async_val":      task["async"],
		"poll":           task["poll"],
		"delegate_to":    task["delegate_to"],
		"delegate_facts": getOr(task, "delegate_facts", false),
		"become":         task["become"],
		"become_user":    task["become_user"],
		"environment":    getOr(task, "environment", map[string]any{}),
		"tags":           getOr(task, "tags", []any{}),
		"notify":         getOr(task, "notify", []any{}),
	}

	// Identifier le module : première clé non réservée, sinon "action".
	moduleKeys := []string{}
	for key := range task {
		if !taskReservedKeys[key] {
			moduleKeys = append(moduleKeys, key)
		}
	}
	sort.Strings(moduleKeys)

	if len(moduleKeys) > 0 {
		parsed["module"] = moduleKeys[0]
		parsed["args"] = getOr(task, moduleKeys[0], map[string]any{})
	} else if action, found := task["action"]; found {
		switch a := action.(type) {
		case map[string]any:
			for key, value := range a {
				parsed["module"] = key
				parsed["args"] = value
				break
			}
		case string:
			if fields := splitFirst(a); fields != "" {
				parsed["module"] = fields
			}
		}
	}

	// Includes, imports, rôles et blocs.
	switch {
	case has(task, "include"):
		parsed["type"] = "include"
		parsed["include"] = task["include"]
	case has(task, "include_tasks"):
		parsed["type"] = "include_tasks"
		parsed["include_tasks"] = task["include_tasks"]
	case has(task, "import_tasks"):
		parsed["type"] = "import_tasks"
		parsed["import_tasks"] = task["import_tasks"]
	case has(task, "include_role"):
		parsed["type"] = "include_role"
		parsed["include_role"] = task["include_role"]
	case has(task, "import_role"):
		parsed["type"] = "import_role"
		parsed["import_role"] = task["import_role"]
	case has(task, "role"):
		parsed["type"] = "role"
		parsed["role"] = task["role"]
	case has(task, "block"):
		parsed["type"] = "block"
		parsed["block"] = parseTasks(task["block"])
		parsed["rescue"] = parseTasks(task["rescue"])
		parsed["always"] = parseTasks(task["always"])
	default:
		parsed["type"] = "task"
	}

	return parsed
}

func parseRoles(raw any) []map[string]any {
	list, ok := raw.([]any)
	if !ok {
		return []map[string]any{}
	}
	parsed := []map[string]any{}
	for _, item := range list {
		switch role := item.(type) {
		case string:
			parsed = append(parsed, map[string]any{"name": role, "vars": map[string]any{}})
		case map[string]any:
			name := str(role["role"])
			if name == "" {
				name = str(role["name"])
			}
			if name == "" {
				name = "unknown"
			}
			vars := map[string]any{}
			for key, value := range role {
				if key != "role" && key != "name" {
					vars[key] = value
				}
			}
			parsed = append(parsed, map[string]any{"name": name, "vars": vars})
		}
	}
	return parsed
}

// AnalyzePlaybook produit l'analyse complète d'un playbook avec statistiques.
func AnalyzePlaybook(content string) map[string]any {
	parsed := ParsePlaybook(content)
	if parsed == nil {
		return map[string]any{"error": "Impossible de parser le playbook"}
	}

	totalTasks, totalHandlers, totalPreTasks, totalPostTasks, totalRoles := 0, 0, 0, 0, 0
	modulesUsed := map[string]bool{}
	hostsTargeted := map[string]bool{}
	collectionsUsed := map[string]bool{}
	becomeUsed := false

	plays, _ := parsed["plays"].([]map[string]any)
	for _, play := range plays {
		tasks := taskList(play["tasks"])
		handlers := taskList(play["handlers"])
		preTasks := taskList(play["pre_tasks"])
		postTasks := taskList(play["post_tasks"])

		totalTasks += len(tasks)
		totalHandlers += len(handlers)
		totalPreTasks += len(preTasks)
		totalPostTasks += len(postTasks)
		if roles, ok := play["roles"].([]map[string]any); ok {
			totalRoles += len(roles)
		}

		switch hosts := play["hosts"].(type) {
		case string:
			hostsTargeted[hosts] = true
		case []any:
			for _, h := range hosts {
				if hs, ok := h.(string); ok {
					hostsTargeted[hs] = true
				}
			}
		}

		if become, ok := play["become"].(bool); ok && become {
			becomeUsed = true
		}
		if collections, ok := play["collections"].([]any); ok {
			for _, c := range collections {
				if cs, ok := c.(string); ok {
					collectionsUsed[cs] = true
				}
			}
		}

		for _, list := range [][]map[string]any{tasks, handlers, preTasks, postTasks} {
			for _, task := range list {
				if module, ok := task["module"].(string); ok && module != "" {
					modulesUsed[module] = true
				}
			}
		}
	}

	return map[string]any{
		"parsed": parsed,
		"statistics": map[string]any{
			"total_plays":      parsed["play_count"],
			"total_tasks":      totalTasks,
			"total_handlers":   totalHandlers,
			"total_pre_tasks":  totalPreTasks,
			"total_post_tasks": totalPostTasks,
			"total_roles":      totalRoles,
			"modules_used":     sortedKeys(modulesUsed),
			"hosts_targeted":   sortedKeys(hostsTargeted),
			"become_used":      becomeUsed,
			"collections_used": sortedKeys(collectionsUsed),
		},
	}
}

// ── Helpers ──────────────────────────────────────────────────────────────────

// toStringMap convertit les maps YAML (clés any) en map[string]any.
func toStringMap(raw any) (map[string]any, bool) {
	if m, ok := raw.(map[string]any); ok {
		return m, true
	}
	return nil, false
}

func getOr(m map[string]any, key string, fallback any) any {
	if v, found := m[key]; found && v != nil {
		return v
	}
	return fallback
}

func has(m map[string]any, key string) bool {
	_, found := m[key]
	return found
}

func taskList(v any) []map[string]any {
	if list, ok := v.([]map[string]any); ok {
		return list
	}
	return nil
}

func sortedKeys(set map[string]bool) []string {
	keys := make([]string, 0, len(set))
	for k := range set {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

func splitFirst(s string) string {
	for i, r := range s {
		if r == ' ' {
			return s[:i]
		}
	}
	return s
}

func str(v any) string {
	s, _ := v.(string)
	return s
}
